File: BinaryTree.py
from time import perf_counter_ns
from random import randint
import sys

class BinaryTree:

    class Node:
        def __init__(self):
            self.id = self.__generateId(25)
            self.left = None
            self.right = None
    
        def __generateId(self, length):
            chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890'
            result = ''
            for i in range(len(chars)):
                result += chars[randint(0, len(chars)-1)]
            return result

    def __init__(self):
        self.root = None
        self.elapsed = 0
        self.count = 0

    def insertNodes(self, nodes):
        print('Start Building Tree with {:d} nodes.'.format(nodes))
        self.__startTimer()
        for i in range(nodes):
            n = BinaryTree.Node()
            self.root = self.__insert(self.root, n)
        duration = self.__stopTimer()
        print('Done Building Tree in {:f} seconds.'.format(duration/1000.0))

    def __insert(self, root, n):
        if root == None:
            root = n
            return root
        else:
            if n.id > root.id:
                root.right = self.__insert(root.right, n)
            else:
                root.left = self.__insert(root.left, n)
            return root

    def walk(self, show):
        self.count = 0
        print('Start Walking Tree.')
        self.__startTimer()
        self.__visit(self.root, show)
        duration = self.__stopTimer()
        print('Done Walking Tree, saw {:d} Nodes in {:f} seconds.'.format(self.count, duration/1000.0))        

    def __visit(self, n, show):
        if n != None:
            self.__visit(n.left, show)
            self.count += 1
            if show:
                print(n.id)
            self.__visit(n.right, show)


    def __startTimer(self):
        self.elapsed = perf_counter_ns()

    def __stopTimer(self):
        return (perf_counter_ns() - self.elapsed)/1000000


nodes = 1000
showNames = False
if len(sys.argv) > 1:
    nodes = int(sys.argv[1])
if len(sys.argv) > 2:
    showNames = True

bt = BinaryTree()
bt.insertNodes(nodes)
bt.walk(showNames)

File: BinaryTree.js
var util = require('util');

class Node {

    constructor() {
        this.id = this.generateId(25);
        this.left = null;
        this.right = null;
    }

    generateId(length) {
        var result = '';
        var chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890';
        var charlen = chars.length
        for (let i=0; i < charlen; i++ ) {
            result += chars.charAt(Math.floor(Math.random() * charlen));
        }
        return result;
    }

}

class BinaryTree {

    constructor() {
        this.root = null;
        this.elapsed = 0;
        this.count = 0;
    }

    insertNodes(nodes) {
        console.log(util.format('Start Building Tree with %d nodes.', nodes));
        this.startTimer();
        for (let i=0; i < nodes; i++) {
            var n = new Node()
            this.root = this.insert(this.root, n);
        }
        var duration = this.stopTimer();
        console.log(util.format('Done Building Tree in %f seconds.', duration/1000.0));
    }

    insert(root, n){
        if ( root == null ) {
            root = n;
            return root;
        } else {
            if ( n.id > root.id ) {
                root.right = this.insert(root.right, n);
            } else {
                root.left = this.insert(root.left, n);
            }
            return root;
        }
    }

    walk(show) {
        this.count = 0;
        console.log('Start Walking Tree.');
        this.startTimer();
        this.visit(this.root, show);
        var duration = this.stopTimer();
        console.log(util.format('Done Walking Tree, saw %d Nodes in %f seconds.', this.count, duration/1000.0));
    }

    visit(n, show) {
        if ( n != null ) {
            this.visit(n.left, show);
            this.count++;
            if (show) {
                console.log(util.format('%s', n.id));
            }
            this.visit(n.right, show);
        }
    }

    startTimer() { this.elapsed = Date.now(); } // already in millis
    stopTimer() { return Date.now() - this.elapsed; }

}

var nodes = 1000000;
var showNames = false;

const theArgs = process.argv.slice(2);  // Remove first two args ('node' and name of this file)
if (theArgs.length > 0) {
    nodes = parseInt(theArgs[0]);
}
if (theArgs.length > 1) {
    showNames = true;
}

var bt = new BinaryTree();
bt.insertNodes(nodes);
bt.walk(showNames);